File: src/main/java/fi/quanfoxes/phases/ConfigurationPhase.java
package fi.quanfoxes.phases;

import java.io.File;
import java.util.ArrayDeque;
import java.util.Arrays;
import java.util.Queue;
import java.util.Vector;

import fi.quanfoxes.BinaryType;
import fi.quanfoxes.Bundle;
import fi.quanfoxes.Phase;
import fi.quanfoxes.Status;

public class ConfigurationPhase extends Phase {

    public static final String EXTENSION = ".z";
    public static final String DEFAULT_OUTPUT = "z";

    private Vector<String> libraries = new Vector<>();
    private Vector<String> files = new Vector<>();

    private void collect(Bundle bundle, File folder, boolean recursive) {
        for (File item : folder.listFiles()) {
            if (item.isFile()) {
                if (item.getName().endsWith(EXTENSION)) {
                    files.add(item.getAbsolutePath());
                }
            }
            else if (recursive) {
                collect(bundle, item, true);
            }
        }
    }

    private Status configure(Bundle bundle, String option, Queue<String> parameters) {
        switch (option) {
            case "-help":
            case "--help": {
                String[][] options = 
                {
                    { "-help / --help",                         "Displays this information" },
                    { "-r <folder> / --recursive <folder>",     "Includes source files (.z) from the given folder and its subfolders"},
                    { "-d / --debug",                           "Generates the output binary with debug info" },
                    { "-o <filename> / --output <filename>",    "Sets the output filename (Default: z.asm, z.o, z, ...)" },
                    { "-l <library> / --library <library>",     "Includes a library to the compilation process" },
                    { "--asm",                                  "Exports the generated assembly to a file" },
                    { "--shared",                               "Sets the output type to shared library (.so)" },
                    { "--static",                               "Sets the output type to static library (.a)"}, 
                    { "-st / --single-thread",                  "Compiles on a single thread instead of multiple threads" },
                    { "-q / --quiet",                           "Suppresses the console output" }
                };

                StringBuilder builder = new StringBuilder();

                for (String[] row : options) {
                    String name = row[0];
                    String description = row[1];

                    builder.append(String.format("\t%-40s%s\n", name, description));
                }

                System.out.println
                (
                    "Usage: zz [options] <folders / files>" + "\n" +
                    "Options:" + "\n" +
                    builder.toString()
                );

                abort();
                return Status.OK;
            }

            case "-r":
            case "--recursive": {
                String folder = parameters.poll();

                if (folder == null || isOption(folder)) {
                    return Status.error("Missing or invalid value for option '%s'", option);
                }

                File handle = new File(folder);

                if (!handle.exists()) {
                    return Status.error("Couldn't find folder '%s'", folder);
                }

                collect(bundle, handle, true);
                return Status.OK;
            }

            case "-d":
            case "--debug": {
                bundle.putBool("debug", true);
                return Status.OK;
            }

            case "-o":
            case "--output": {
                String output = parameters.poll();

                if (output == null || isOption(output)) {
                    return Status.error("Missing or invalid value for option '%s'", option);
                }

                bundle.put("output", output);
                return Status.OK;
            }

            case "-l":
            case "--lib":
            case "--library": {
                String library = parameters.poll();

                if (library == null || isOption(library)) {
                    return Status.error("Missing or invalid value for option '%s'", option);
                }

                libraries.add(library);
                return Status.OK;
            }

            case "--asm": {
                bundle.putBool("assembly", true);
                return Status.OK;
            }

            case "--shared": {
                bundle.put("output_type", BinaryType.SHARED_LIBRARY);
                return Status.OK;
            }

            case "--static": {
                bundle.put("output_type", BinaryType.STATIC_LIBRARY);
                return Status.OK;
            }

            case "-st":
            case "--single-thread": {
                bundle.putBool("multithreaded", false);
                return Status.OK;
            }

            default: {
                return Status.error("Unknown option");
            }
        }
    }

    private boolean isOption(String element) {
        return element.charAt(0) == '-';
    }

    @Override
    public Status execute(Bundle bundle) {
        String[] arguments = bundle.get("arguments", null);

        if (arguments == null) {
            return Status.error("Couldn't configure settings");
        }

        Queue<String> parameters = new ArrayDeque<String>(Arrays.asList(arguments));

        while (!parameters.isEmpty()) {
            String element = parameters.poll();

            if (isOption(element)) {
                Status status = configure(bundle, element, parameters);

                if (status.isProblematic()) {
                    return status;
                }
            }
            else {
                File file = new File(element);

                if (!file.exists()) {
                    return Status.error("Invalid source file/folder '%s'", element);
                }

                if (file.isDirectory()) {
                    collect(bundle, file.getAbsoluteFile(), false);
                }
                else if (file.getName().endsWith(EXTENSION)) {
                    files.add(file.getAbsolutePath());
                }
                else {
                    return Status.error("Source file must have '%s' extension", EXTENSION);
                }
            }
        }

        bundle.put("input_files", files.toArray(new String[files.size()]));
        bundle.put("libraries", libraries.toArray(new String[libraries.size()]));

        if (!bundle.has("output")) {
            bundle.put("output", DEFAULT_OUTPUT);
        }

        return Status.OK;
    }
}

File: Zigzag/Phases/ConfigurationPhase.cs
using System;
using System.Collections.Generic;
using System.IO;

public class ConfigurationPhase : Phase
{
	public const string EXTENSION = ".z";
	public const string DEFAULT_OUTPUT = "z";

	private List<string> Libraries { get; set; } = new List<string>();
	private List<string> Files { get; set; } = new List<string>();

	private void Collect(Bundle bundle, DirectoryInfo folder, bool recursive = true)
	{
		foreach (FileInfo item in folder.GetFiles())
		{
			if (item.Extension == EXTENSION)
			{
				Files.Add(item.FullName);
			}
		}

		if (recursive)
		{
			foreach (DirectoryInfo item in folder.GetDirectories())
			{
				Collect(bundle, item);
			}
		}
	}

	private struct Option
	{
		public string Command { get; set; }
		public string Description { get; set; }

		public override string ToString()
		{
			return $"\t{Command,-40}{Description}\n";
		}
	};

	private Status Configure(Bundle bundle, string option, Queue<string> parameters)
	{
		switch (option)
		{
			case "-help":
			case "--help":
			{
				Option[] options = new Option[]
				{
					new Option() { Command = "-help / --help",                         Description = "Displays this information" },
					new Option() { Command = "-r <folder> / --recursive <folder>",     Description = "Includes source files (.z) from the given folder and its subfolders"},
					new Option() { Command = "-d / --debug",                           Description = "Generates the output binary with debug info" },
					new Option() { Command = "-o <filename> / --output <filename>",    Description = "Sets the output filename (Default: z.asm, z.o, z, ...)" },
					new Option() { Command = "-l <library> / --library <library>",     Description = "Includes a library to the compilation process" },
					new Option() { Command = "--asm",                                  Description = "Exports the generated assembly to a file" },
					new Option() { Command = "--shared",                               Description = "Sets the output type to shared library (.so)" },
					new Option() { Command = "--static",                               Description = "Sets the output type to static library (.a)"},
					new Option() { Command = "-st / --single-thread",                  Description = "Compiles on a single thread instead of multiple threads" },
					new Option() { Command = "-q / --quiet",                           Description = "Suppresses the console output" }
				};

				Console.WriteLine
				(
					"Usage: zz [options] <folders / files>" + "\n" +
					"Options:" + "\n" +
					string.Join(null, options)
				);

				Abort();

				return Status.OK;
			}

			case "-r":
			case "--recursive":
			{
				string folder = parameters.Dequeue();

				if (folder == null || IsOption(folder))
				{
					return Status.Error("Missing or invalid value for option '{0}'", option);
				}

				if (!Directory.Exists(folder))
				{
					return Status.Error("Couldn't find folder '{0}'", folder);
				}

				Collect(bundle, new DirectoryInfo(folder), true);

				return Status.OK;
			}

			case "-d":
			case "--debug":
			{
				bundle.PutBool("debug", true);
				return Status.OK;
			}

			case "-o":
			case "--output":
			{
				string output = parameters.Dequeue();

				if (output == null || IsOption(output))
				{
					return Status.Error("Missing or invalid value for option '{0}'", option);
				}

				bundle.Put("output", output);
				return Status.OK;
			}

			case "-l":
			case "--lib":
			case "--library":
			{
				string library = parameters.Dequeue();

				if (library == null || IsOption(library))
				{
					return Status.Error("Missing or invalid value for option '{0}'", option);
				}

				Libraries.Add(library);
				return Status.OK;
			}

			case "--asm":
			{
				bundle.PutBool("assembly", true);
				return Status.OK;
			}

			case "--shared":
			{
				bundle.Put("output_type", BinaryType.SHARED_LIBRARY);
				return Status.OK;
			}

			case "--static":
			{
				bundle.Put("output_type", BinaryType.STATIC_LIBRARY);
				return Status.OK;
			}

			case "-st":
			case "--single-thread":
			{
				bundle.PutBool("multithreaded", false);
				return Status.OK;
			}

			case "--time":
			{
				bundle.PutBool("time", true);
				return Status.OK;
			}

			default:
			{
				return Status.Error("Unknown option");
			}
		}
	}

	private bool IsOption(string element)
	{
		return element[0] == '-';
	}

	public override Status Execute(Bundle bundle)
	{
		string[] arguments = bundle.Get<string[]>("arguments", null);

		if (arguments == null)
		{
			return Status.Error("Couldn't configure settings");
		}

		Queue<string> parameters = new Queue<string>(arguments);

		while (parameters.Count > 0)
		{
			string element = parameters.Dequeue();

			if (IsOption(element))
			{
				Status status = Configure(bundle, element, parameters);

				if (status.IsProblematic)
				{
					return status;
				}
			}
			else
			{
				FileInfo file = new FileInfo(element);

				if (file.Exists)
				{
					if (file.Extension == EXTENSION)
					{
						Files.Add(file.FullName);
						continue;
					}
					else
					{
						return Status.Error($"Source file must have '{EXTENSION}' extension");
					}
				}

				DirectoryInfo directory = new DirectoryInfo(element);

				if (directory.Exists)
				{
					Collect(bundle, directory, false);
					continue;
				}

				return Status.Error("Invalid source file/folder '{0}'", element);
			}
		}

		bundle.Put("input_files", Files.ToArray());
		bundle.Put("libraries", Libraries.ToArray());

		if (!bundle.Contains("output"))
		{
			bundle.Put("output", DEFAULT_OUTPUT);
		}

		return Status.OK;
	}
}